### PR TITLE
Add support for the `SetupIntent` resource and APIs

### PR DESCRIPTION
**Pull request by remi-stripe**
r? @brandur-stripe 
cc @stripe/api-libraries

### CODE REVIEW

File: paymentintent.go [hunk 1]
```diff
@@ -56,6 +56,15 @@ const (
 	PaymentIntentOffSessionRecurring PaymentIntentOffSession = "recurring"
 )
 
+// PaymentIntentSetupFutureUsage is the list of allowed values for SetupFutureUsage.
+type PaymentIntentSetupFutureUsage string
+
+// List of values that PaymentIntentSetupFutureUsage can take.
+const (
+	PaymentIntentSetupFutureUsageOffSession PaymentIntentSetupFutureUsage = "off_session"
+	PaymentIntentSetupFutureUsageOnSession  PaymentIntentSetupFutureUsage = "on_session"
```
Comment (brandur-stripe) [inline paymentintent.go]: I'm sure you're well aware, but: It looks like this is still correct according to current docs, but I thought we'd decided to rename in the end.

File: setupintent.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	LastSetupError     *Error                   `json:"last_setup_error"`
+	Livemode           bool                     `json:"livemode"`
+	Metadata           map[string]string        `json:"metadata"`
+	NextAction         *PaymentIntentNextAction `json:"next_action"`
+	Object             string                   `json:"object"`
+	OnBehalfOf         *Account                 `json:"on_behalf_of"`
+	PaymentMethod      *PaymentMethod           `json:"payment_method"`
+	PaymentMethodTypes []string                 `json:"payment_method_types"`
+	Status             SetupIntentStatus        `json:"status"`
+	Usage              SetupIntentUsage         `json:"usage"`
```
Comment (brandur-stripe) [inline setupintent.go]: Just flagging that you included `usage` which is marked as a preview feature, but omitted `requested_credentials` and `single_use`. Not sure if that was intentional or not.

File: setupintent.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+// UnmarshalJSON handles deserialization of a SetupIntent.
+// This custom unmarshaling is needed because the resulting
+// property may be an id or the full struct if it was expanded.
+func (p *SetupIntent) UnmarshalJSON(data []byte) error {
+	if id, ok := ParseID(data); ok {
+		p.ID = id
+		return nil
+	}
+
+	type paymentintent SetupIntent
```
Comment (brandur-stripe) [inline setupintent.go]: Maybe rename this to `setupIntent` (I'd use a capitalized "I" in "Intent" probably).

File: paymentintent.go (hunk 1 above)
Comment (remi-stripe) [reply paymentintent.go]: Yes, I thought so too but after all we did not. They gavel-ed the original proposal!

Comment (brandur-stripe) [reply paymentintent.go]: Ah, weird. SG though.

File: setupintent.go (hunk 2 above)
Comment (remi-stripe) [reply setupintent.go]: Confirmed internally that this one is gated by mistake and will ship.

File: setupintent.go (hunk 3 above)
Comment (remi-stripe) [reply setupintent.go]: fixed

**Comment by remi-stripe** [comment]
@brandur-stripe thanks for catching the issues. PTAL

**Comment by remi-stripe** [comment]
self-approving as I just added a new field `cancellation_reason`.

**Comment by brandur-stripe** [comment]
We found some bad logic in this one — reverted in ec2279d and removed from the 61.8.0 release.

**Pull request merged**